File: common/src/sbi.cpp
#include "mabur/sbi.h"
#include "mabur/crc16.h"

namespace mabur {

SbiPacker::SbiPacker(int block_payload, int blocks_per_body, uint8_t stream_id)
    : block_payload_(block_payload),
      blocks_per_body_(blocks_per_body),
      stream_id_(stream_id) {}

int SbiPacker::block_stride() const { return 2 + block_payload_; }

std::vector<std::vector<uint8_t>> SbiPacker::add(const uint8_t* env, size_t len) {
  std::vector<std::vector<uint8_t>> out;
  auto b = add_one(env, len);
  if (!b.empty()) out.push_back(std::move(b));
  return out;
}

std::vector<std::vector<uint8_t>> SbiPacker::flush() {
  std::vector<std::vector<uint8_t>> out;
  auto b = flush_one();
  if (!b.empty()) out.push_back(std::move(b));
  return out;
}

void SbiPacker::begin_body() {
  body_.clear();
  body_.reserve(static_cast<size_t>(SBI_HDR_LEN) +
                static_cast<size_t>(blocks_per_body_) * static_cast<size_t>(block_stride()));
  // Header: <u16 MAGIC LE, u8 ver, u8 stream_id, u16 block_payload LE, u8 n_blocks, u16 q_ms LE, u16 enc_us LE, u16 air_ms LE>
  body_.push_back(static_cast<uint8_t>(SBI_MAGIC & 0xFF));
  body_.push_back(static_cast<uint8_t>((SBI_MAGIC >> 8) & 0xFF));
  body_.push_back(SBI_VER);
  body_.push_back(stream_id_);
  body_.push_back(static_cast<uint8_t>(block_payload_ & 0xFF));
  body_.push_back(static_cast<uint8_t>((block_payload_ >> 8) & 0xFF));
  body_.push_back(0);  // n_blocks, patched by take_body
  // q_ms (7-8), enc_us (9-10), air_ms (11-12) placeholders, patched later
  // by the tx thread / the hot thread's sink
  for (int i = 0; i < 6; ++i) body_.push_back(0);
}

std::vector<uint8_t> SbiPacker::take_body() {
  body_[6] = static_cast<uint8_t>(n_pending_);
  n_pending_ = 0;
  std::vector<uint8_t> out = std::move(body_);
  body_.clear();
  return out;
}

std::vector<uint8_t> SbiPacker::add_one(const uint8_t* env, size_t len) {
  if (static_cast<int>(len) != block_payload_) return {};
  if (n_pending_ == 0) begin_body();
  const uint16_t crc = crc16_ccitt(env, len);
  body_.push_back(static_cast<uint8_t>(crc & 0xFF));
  body_.push_back(static_cast<uint8_t>((crc >> 8) & 0xFF));
  body_.insert(body_.end(), env, env + len);
  if (++n_pending_ >= blocks_per_body_) return take_body();
  return {};
}

std::vector<uint8_t> SbiPacker::flush_one() {
  if (n_pending_ == 0) return {};
  return take_body();
}

namespace {
uint16_t sbi_rd_u16(const uint8_t* p) {
  return static_cast<uint16_t>(p[0] | (p[1] << 8));
}
}  // namespace

SbiUnpackResult sbi_unpack(const uint8_t* body, size_t len, int block_payload) {
  SbiUnpackResult r;
  // Memory-safety guard: negative block_payload would wrap stride to a huge value,
  // causing unbounded reads in crc16_ccitt(). Python is safe via slice semantics.
  if (block_payload <= 0) return r;
  const size_t stride = 2 + static_cast<size_t>(block_payload);
  if (len >= static_cast<size_t>(SBI_HDR_LEN)) {
    const uint16_t magic = sbi_rd_u16(body);
    const uint8_t ver = body[2];
    r.stream_id = body[3];
    const uint16_t hdr_bp = sbi_rd_u16(body + 4);
    r.header_ok = magic == SBI_MAGIC && ver == SBI_VER && hdr_bp == block_payload;
    if (r.header_ok) {
      r.q_ms = sbi_rd_u16(body + SBI_Q_MS_OFF);
      r.enc_us = sbi_rd_u16(body + SBI_ENC_US_OFF);
      r.air_ms = sbi_rd_u16(body + SBI_AIR_MS_OFF);
    }
  } else {
    return r;  // Python: empty region -> zero blocks, header_ok false
  }
  const uint8_t* region = body + SBI_HDR_LEN;
  const size_t region_len = len - static_cast<size_t>(SBI_HDR_LEN);
  r.n_blocks = static_cast<int>(region_len / stride);
  for (int i = 0; i < r.n_blocks; ++i) {
    const uint8_t* off = region + static_cast<size_t>(i) * stride;
    const uint16_t crc_field = sbi_rd_u16(off);
    if (crc16_ccitt(off + 2, static_cast<size_t>(block_payload)) == crc_field)
      r.survivors.emplace_back(off + 2, off + 2 + block_payload);
    else
      ++r.n_failed;
  }
  return r;
}

int sbi_peek_stream_id(const uint8_t* body, size_t len) {
  if (len < static_cast<size_t>(SBI_HDR_LEN)) return -1;
  if (sbi_rd_u16(body) != SBI_MAGIC || body[2] != SBI_VER) return -1;
  return body[3];
}

void sbi_set_q_ms(uint8_t* body, size_t len, uint16_t ms) {
  if (len < static_cast<size_t>(SBI_HDR_LEN)) return;
  body[SBI_Q_MS_OFF] = static_cast<uint8_t>(ms & 0xFF);
  body[SBI_Q_MS_OFF + 1] = static_cast<uint8_t>(ms >> 8);
}

void sbi_set_enc_us(uint8_t* body, size_t len, uint16_t us) {
  if (len < static_cast<size_t>(SBI_HDR_LEN)) return;
  body[SBI_ENC_US_OFF] = static_cast<uint8_t>(us & 0xFF);
  body[SBI_ENC_US_OFF + 1] = static_cast<uint8_t>(us >> 8);
}

void sbi_set_air_ms(uint8_t* body, size_t len, uint16_t ms) {
  if (len < static_cast<size_t>(SBI_HDR_LEN)) return;
  body[SBI_AIR_MS_OFF] = static_cast<uint8_t>(ms & 0xFF);
  body[SBI_AIR_MS_OFF + 1] = static_cast<uint8_t>(ms >> 8);
}

}  // namespace mabur

File: tools/pyref/sbi.py
"""SBI (Sub-Block Integrity) ver 1 packer for test vector generation.
Override of devourer's fec_subblock for latency-accounting: adds q_ms and enc_us duration fields.
Ver 0 is hard-rejected (flag-day deploy, no compatibility).
"""
import struct
from typing import List

SBI_MAGIC = 0xF5B0
SBI_VER = 1
SBI_HDR_STRUCT = "<HBBHBHH"  # MAGIC, VER, STREAM_ID, BLOCK_PAYLOAD, N_BLOCKS, Q_MS, ENC_US
SBI_HDR_LEN = struct.calcsize(SBI_HDR_STRUCT)  # 11 bytes
SBI_Q_MS_OFF = 7
SBI_ENC_US_OFF = 9


class SubBlockPacker:
    """Ver 1 SBI packer: 11-byte header with q_ms and enc_us latency fields.
    Replaces devourer's version for test vector generation only."""

    def __init__(self, block_payload: int, blocks_per_body: int, stream_id: int = 0):
        self.block_payload = block_payload
        self.blocks_per_body = blocks_per_body
        self.stream_id = stream_id
        self.pending: List[bytes] = []

    def block_stride(self) -> int:
        """Per-block wire size: crc16 (2 bytes) + block_payload."""
        return 2 + self.block_payload

    def add(self, env: bytes) -> List[bytes]:
        """Add one FEC envelope. Returns completed bodies when blocks_per_body accumulate."""
        if len(env) != self.block_payload:
            return []
        self.pending.append(bytes(env))
        out = []
        while len(self.pending) >= self.blocks_per_body:
            batch = self.pending[: self.blocks_per_body]
            self.pending = self.pending[self.blocks_per_body :]
            out.append(self._build_body(batch))
        return out

    def flush(self) -> List[bytes]:
        """Emit final body with pending envelopes. Returns empty if nothing pending."""
        if not self.pending:
            return []
        out = self._build_body(self.pending)
        self.pending = []
        return [out]

    def _build_body(self, batch: List[bytes]) -> bytes:
        """Build one SBI body from a batch of envelopes."""
        out = bytearray()

        # Header: <u16 MAGIC LE, u8 ver, u8 stream_id, u16 block_payload LE, u8 n_blocks, u16 q_ms LE, u16 enc_us LE>
        # q_ms and enc_us are initialized as zero placeholders.
        out.extend(
            struct.pack(
                SBI_HDR_STRUCT,
                SBI_MAGIC,
                SBI_VER,
                self.stream_id,
                self.block_payload,
                len(batch),
                0,  # q_ms placeholder
                0,  # enc_us placeholder
            )
        )

        # Append sub-blocks: <u16 crc16 LE, payload>
        for env in batch:
            crc = self._crc16_ccitt(env)
            out.extend(struct.pack("<H", crc))
            out.extend(env)

        return bytes(out)

    @staticmethod
    def _crc16_ccitt(data: bytes) -> int:
        """CRC16-CCITT (same as devourer's implementation)."""
        crc = 0xFFFF
        for byte in data:
            crc ^= byte << 8
            for _ in range(8):
                crc <<= 1
                if crc & 0x10000:
                    crc ^= 0x1021
                crc &= 0xFFFF
        return crc


def unpack(body: bytes, block_payload: int) -> dict:
    """Unpack an SBI body into surviving sub-blocks (ver 1).
    Returns: {survivors: list of valid payloads, n_blocks, n_failed, header_ok, stream_id, q_ms, enc_us}
    """
    result = {
        "survivors": [],
        "n_blocks": 0,
        "n_failed": 0,
        "header_ok": False,
        "stream_id": 0,
        "q_ms": 0,
        "enc_us": 0,
    }

    if block_payload <= 0 or len(body) < SBI_HDR_LEN:
        return result

    magic, ver, stream_id, hdr_bp, n_blocks, q_ms, enc_us = struct.unpack_from(
        SBI_HDR_STRUCT, body
    )

    result["stream_id"] = stream_id
    result["header_ok"] = (
        magic == SBI_MAGIC and ver == SBI_VER and hdr_bp == block_payload
    )
    if result["header_ok"]:
        result["q_ms"] = q_ms
        result["enc_us"] = enc_us

    region = body[SBI_HDR_LEN :]
    stride = 2 + block_payload
    result["n_blocks"] = len(region) // stride

    for i in range(result["n_blocks"]):
        off = i * stride
        crc_field = struct.unpack_from("<H", region[off : off + 2])[0]
        payload = region[off + 2 : off + 2 + block_payload]
        if len(payload) == block_payload:
            computed_crc = SubBlockPacker._crc16_ccitt(payload)
            if crc_field == computed_crc:
                result["survivors"].append(payload)
            else:
                result["n_failed"] += 1

    return result


def peek_stream_id(body: bytes) -> int:
    """Peek stream_id from SBI header (ver 1), or -1 on invalid header."""
    if len(body) < SBI_HDR_LEN:
        return -1
    magic, ver, stream_id, _, _, _, _ = struct.unpack_from(SBI_HDR_STRUCT, body)
    if magic != SBI_MAGIC or ver != SBI_VER:
        return -1
    return stream_id
